dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +253,282 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+        return isTaskRegistered;
+    }
 
-            if (isProtocolActivatorSet)
+    void PushNotificationManager::Register()
+    {
+        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+
+        try
+        {
+            if (!AppModel::Identity::IsPackagedProcess())
             {
                 {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_protocolRegistration);
+                    auto lock = m_lock.lock_shared();
+                    THROW_HR_IF(E_INVALIDARG, m_protocolRegistration);
                 }
 
                 // AppId generated by PushNotificationsLongRunningTask singleton
                 wil::unique_cotaskmem_string unpackagedAppUserModelId;
                 RegisterUnpackagedApplicationHelper(GUID_NULL, unpackagedAppUserModelId); // create default registration for app
 
                 {
-                    auto lock = s_activatorInfoLock.lock_exclusive();
-                    s_protocolRegistration = true;
+                    auto lock = m_lock.lock_exclusive();
+                    // Flag protects from multiple registrations
+                    m_protocolRegistration = true;
                 }
             }
 
-            BackgroundTaskBuilder builder{ nullptr };
-
-            if (WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::PushTrigger) && PushNotificationHelpers::IsPackagedAppScenario())
+            winrt::guid registeredClsid{ GUID_NULL };
+            if (AppModel::Identity::IsPackagedProcess())
             {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
+                // Returns ComActivator CLSID from registry. This CLSID provided in manifest is registered when a packaged app is installed
+                THROW_IF_FAILED(PushNotificationHelpers::GetComRegistrationFromRegistry(expectedPushServerArgs.data(), registeredClsid));
+            }
 
-                {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_pushTriggerRegistration);
-                }
+            auto scopeExitToCleanRegistrations{ wil::scope_exit(
+            [&](){
+                    auto lock{ m_lock.lock_exclusive() };
 
-                winrt::hstring taskClsidStr{ winrt::to_hstring(taskClsid) };
-                winrt::hstring backgroundTaskFullName{ backgroundTaskName + taskClsidStr };
+                    m_comActivatorRegistration.reset();
 
-                auto tasks{ BackgroundTaskRegistration::AllTasks() };
-                bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
-                    [&](auto&& task)
+                    // Clean the task registration only if it was created during this call
+                    if (m_pushTriggerRegistration)
```

Comment:
What about cleanup of the Sink with the platform or LongRunning singleton if registration fails?

Reply:
@pmpurifoy